File: ros2-bridge/rosserial2/codec.py
"""Wire-frame codec for rosserial2 protocol v1.

Frame layout (little-endian throughout):

    0xAA 0x55  LEN_LO LEN_HI  SEQ  MSG_ID  PAYLOAD...  CRC32

CRC32 is IEEE 802.3, computed over LEN_LO..end-of-payload. The codec is
pure: no I/O, no globals, no allocation beyond the bytes objects it
returns. A streaming parser is provided for byte-at-a-time use.
"""

from __future__ import annotations

import zlib
from collections.abc import Iterable, Iterator
from dataclasses import dataclass

SYNC: bytes = b"\xaa\x55"
HEADER_SIZE: int = 6  # SYNC(2) + LEN(2) + SEQ(1) + MSG_ID(1)
CRC_SIZE: int = 4
OVERHEAD: int = HEADER_SIZE + CRC_SIZE  # 10
MAX_PAYLOAD: int = 512
CONTROL_MSG_ID: int = 0


class FrameError(ValueError):
    """Raised when a frame fails validation during decode."""


@dataclass(frozen=True)
class Frame:
    """A decoded frame. ``msg_id == 0`` means control plane."""

    seq: int
    msg_id: int
    payload: bytes

    def __post_init__(self) -> None:
        if not 0 <= self.seq <= 0xFF:
            raise FrameError(f"seq out of range: {self.seq}")
        if not 0 <= self.msg_id <= 0xFF:
            raise FrameError(f"msg_id out of range: {self.msg_id}")
        if len(self.payload) > MAX_PAYLOAD:
            raise FrameError(f"payload too large: {len(self.payload)} > {MAX_PAYLOAD}")


def _crc32(data: bytes) -> int:
    return zlib.crc32(data) & 0xFFFFFFFF


def encode_frame(seq: int, msg_id: int, payload: bytes) -> bytes:
    """Encode a single frame to bytes. Validates inputs."""
    if not 0 <= seq <= 0xFF:
        raise FrameError(f"seq out of range: {seq}")
    if not 0 <= msg_id <= 0xFF:
        raise FrameError(f"msg_id out of range: {msg_id}")
    if len(payload) > MAX_PAYLOAD:
        raise FrameError(f"payload too large: {len(payload)} > {MAX_PAYLOAD}")

    length = len(payload)
    crc_region = bytes([length & 0xFF, (length >> 8) & 0xFF, seq, msg_id]) + payload
    crc = _crc32(crc_region)
    return SYNC + crc_region + crc.to_bytes(4, "little")


def decode_frame(buf: bytes) -> Frame:
    """Decode a single complete frame from ``buf``.

    The buffer must contain exactly one frame and nothing else. For
    streaming data, use :class:`FrameParser`.
    """
    if len(buf) < OVERHEAD:
        raise FrameError(f"buffer too short: {len(buf)} < {OVERHEAD}")
    if buf[0:2] != SYNC:
        raise FrameError("bad sync")
    length = buf[2] | (buf[3] << 8)
    if length > MAX_PAYLOAD:
        raise FrameError(f"length out of range: {length}")
    expected_total = OVERHEAD + length
    if len(buf) != expected_total:
        raise FrameError(f"size mismatch: got {len(buf)}, expected {expected_total}")
    seq = buf[4]
    msg_id = buf[5]
    payload = buf[HEADER_SIZE : HEADER_SIZE + length]
    crc_offset = HEADER_SIZE + length
    crc_actual = int.from_bytes(buf[crc_offset : crc_offset + 4], "little")
    crc_expected = _crc32(buf[2:crc_offset])
    if crc_actual != crc_expected:
        raise FrameError(f"crc mismatch: got {crc_actual:#010x}, expected {crc_expected:#010x}")
    return Frame(seq=seq, msg_id=msg_id, payload=bytes(payload))


class FrameParser:
    """Streaming parser. Feed bytes; collect complete frames.

    The parser implements the resync rule from ADR-0002: on any failure
    (bad length, CRC mismatch) it returns to scanning, and resumes from
    the byte *after* the first sync byte of the failed candidate. A
    payload that happens to contain ``0xAA 0x55`` cannot deadlock the
    parser because each candidate either fails the LEN bound or the
    CRC, and the scanner walks past it.

    Implementation: we keep a byte buffer of all unconsumed input. On
    each ``feed()`` we run the scanner until we either produce a frame
    or run out of bytes to make progress. On failure we discard the
    leading 0xAA and continue.
    """

    def __init__(self) -> None:
        self._buf = bytearray()
        # Telemetry.
        self.errors: int = 0
        self.bytes_seen: int = 0

    def reset(self) -> None:
        self._buf.clear()

    def feed(self, data: bytes | bytearray | memoryview) -> list[Frame]:
        """Feed bytes; return any complete frames produced."""
        return list(self.feed_iter(data))

    def feed_iter(self, data: bytes | bytearray | memoryview) -> Iterator[Frame]:
        self.bytes_seen += len(data)
        self._buf.extend(data)
        while True:
            frame, advance = self._try_one()
            if advance == 0:
                return
            del self._buf[:advance]
            if frame is not None:
                yield frame

    def _try_one(self) -> tuple[Frame | None, int]:
        """Try to extract one frame from the head of ``_buf``.

        Returns ``(frame, advance)``:
        - ``advance == 0`` means we need more data; nothing was consumed.
        - ``advance > 0`` means consume that many bytes from the head;
          ``frame`` is the produced frame or ``None`` if those bytes
          were garbage (no sync found, or a failed candidate's leading
          ``0xAA``).
        """
        buf = self._buf
        n = len(buf)
        if n == 0:
            return None, 0

        # Scan for sync.
        try:
            start = buf.index(0xAA)
        except ValueError:
            # No 0xAA at all: drop everything (none of it can begin a
            # frame). Need at least one byte to have entered here.
            return None, n

        if start > 0:
            return None, start

        # buf[0] == 0xAA. Need at least 2 bytes to test the second sync.
        if n < 2:
            return None, 0
        if buf[1] != 0x55:
            # Failed sync: drop just the 0xAA and let the scanner retry.
            return None, 1

        # Need full header to read length.
        if n < HEADER_SIZE:
            return None, 0
        length = buf[2] | (buf[3] << 8)
        if length > MAX_PAYLOAD:
            self.errors += 1
            return None, 1  # rewind past the 0xAA
        total = OVERHEAD + length
        if n < total:
            return None, 0

        seq = buf[4]
        msg_id = buf[5]
        payload = bytes(buf[HEADER_SIZE : HEADER_SIZE + length])
        crc_offset = HEADER_SIZE + length
        crc_actual = int.from_bytes(buf[crc_offset : crc_offset + 4], "little")
        crc_expected = _crc32(bytes(buf[2:crc_offset]))
        if crc_actual != crc_expected:
            self.errors += 1
            return None, 1  # rewind past the 0xAA

        frame = Frame(seq=seq, msg_id=msg_id, payload=payload)
        return frame, total


def chunks(data: Iterable[int], size: int) -> Iterator[bytes]:
    """Utility: split ``data`` into ``size``-byte chunks. Used in tests."""
    buf: list[int] = []
    for b in data:
        buf.append(b)
        if len(buf) == size:
            yield bytes(buf)
            buf.clear()
    if buf:
        yield bytes(buf)

File: web-flasher/js/codec.js
// rosserial2 wire codec — ES module, browser- and Node-compatible.
//
// Mirrors ros2-bridge/rosserial2/codec.py exactly. Same constants,
// same frame layout, same one-byte-rewind streaming parser. The
// cross-language corpus checked into
// ros2-bridge/tests/corpus/ keeps the three implementations in sync.

export const SYNC = Uint8Array.of(0xaa, 0x55);
export const HEADER_SIZE = 6;       // SYNC(2)+LEN(2)+SEQ(1)+MSG_ID(1)
export const CRC_SIZE = 4;
export const OVERHEAD = HEADER_SIZE + CRC_SIZE; // 10
export const MAX_PAYLOAD = 512;
export const CONTROL_MSG_ID = 0;
export const PROTOCOL_VERSION = 1;

export class FrameError extends Error {}

/** IEEE 802.3 CRC32 (poly 0xEDB88320), bitwise. Returns u32. */
export function crc32(bytes) {
    let crc = 0xffffffff >>> 0;
    for (let i = 0; i < bytes.length; i++) {
        crc = (crc ^ bytes[i]) >>> 0;
        for (let b = 0; b < 8; b++) {
            const mask = -(crc & 1) >>> 0;
            crc = ((crc >>> 1) ^ (0xedb88320 & mask)) >>> 0;
        }
    }
    return (~crc) >>> 0;
}

function checkRange(name, value, lo, hi) {
    if (!Number.isInteger(value) || value < lo || value > hi) {
        throw new FrameError(`${name} out of range: ${value}`);
    }
}

/**
 * Encode a frame.
 * @param {number} seq     0..255
 * @param {number} msgId   0..255
 * @param {Uint8Array} payload  length <= MAX_PAYLOAD
 * @returns {Uint8Array}
 */
export function encodeFrame(seq, msgId, payload) {
    checkRange("seq", seq, 0, 0xff);
    checkRange("msg_id", msgId, 0, 0xff);
    if (!(payload instanceof Uint8Array)) {
        throw new FrameError("payload must be a Uint8Array");
    }
    if (payload.length > MAX_PAYLOAD) {
        throw new FrameError(`payload too large: ${payload.length} > ${MAX_PAYLOAD}`);
    }
    const length = payload.length;
    const total = OVERHEAD + length;
    const out = new Uint8Array(total);
    out[0] = 0xaa;
    out[1] = 0x55;
    out[2] = length & 0xff;
    out[3] = (length >>> 8) & 0xff;
    out[4] = seq;
    out[5] = msgId;
    out.set(payload, HEADER_SIZE);
    const crc = crc32(out.subarray(2, HEADER_SIZE + length));
    const crcOffset = HEADER_SIZE + length;
    out[crcOffset + 0] = crc & 0xff;
    out[crcOffset + 1] = (crc >>> 8) & 0xff;
    out[crcOffset + 2] = (crc >>> 16) & 0xff;
    out[crcOffset + 3] = (crc >>> 24) & 0xff;
    return out;
}

/**
 * Decode exactly one frame.
 * @param {Uint8Array} buf
 * @returns {{seq:number, msgId:number, payload:Uint8Array}}
 */
export function decodeFrame(buf) {
    if (!(buf instanceof Uint8Array)) {
        throw new FrameError("buf must be a Uint8Array");
    }
    if (buf.length < OVERHEAD) {
        throw new FrameError(`buffer too short: ${buf.length} < ${OVERHEAD}`);
    }
    if (buf[0] !== 0xaa || buf[1] !== 0x55) {
        throw new FrameError("bad sync");
    }
    const length = buf[2] | (buf[3] << 8);
    if (length > MAX_PAYLOAD) {
        throw new FrameError(`length out of range: ${length}`);
    }
    const total = OVERHEAD + length;
    if (buf.length !== total) {
        throw new FrameError(`size mismatch: got ${buf.length}, expected ${total}`);
    }
    const seq = buf[4];
    const msgId = buf[5];
    const payload = buf.slice(HEADER_SIZE, HEADER_SIZE + length);
    const crcOffset = HEADER_SIZE + length;
    const crcActual =
        (buf[crcOffset] |
            (buf[crcOffset + 1] << 8) |
            (buf[crcOffset + 2] << 16) |
            (buf[crcOffset + 3] << 24)) >>> 0;
    const crcExpected = crc32(buf.subarray(2, crcOffset));
    if (crcActual !== crcExpected) {
        throw new FrameError(
            `crc mismatch: got 0x${crcActual.toString(16)}, expected 0x${crcExpected.toString(16)}`
        );
    }
    return { seq, msgId, payload };
}

/**
 * Streaming parser. ``feed(bytes)`` returns any complete frames
 * produced. Mirrors the Python parser's one-byte-rewind rule.
 */
export class FrameParser {
    constructor() {
        this._buf = new Uint8Array(0);
        this.errors = 0;
        this.bytesSeen = 0;
    }

    reset() {
        this._buf = new Uint8Array(0);
    }

    feed(data) {
        if (!(data instanceof Uint8Array)) {
            throw new FrameError("feed() requires Uint8Array");
        }
        this.bytesSeen += data.length;
        // Concatenate.
        const merged = new Uint8Array(this._buf.length + data.length);
        merged.set(this._buf, 0);
        merged.set(data, this._buf.length);
        this._buf = merged;

        const out = [];
        while (true) {
            const { frame, advance } = this._tryOne();
            if (advance === 0) break;
            this._buf = this._buf.subarray(advance);
            if (frame !== null) out.push(frame);
        }
        return out;
    }

    _tryOne() {
        const buf = this._buf;
        if (buf.length === 0) return { frame: null, advance: 0 };
        let start = 0;
        while (start < buf.length && buf[start] !== 0xaa) start++;
        if (start > 0) return { frame: null, advance: start };
        if (buf.length < 2) return { frame: null, advance: 0 };
        if (buf[1] !== 0x55) return { frame: null, advance: 1 };
        if (buf.length < HEADER_SIZE) return { frame: null, advance: 0 };
        const length = buf[2] | (buf[3] << 8);
        if (length > MAX_PAYLOAD) {
            this.errors++;
            return { frame: null, advance: 1 };
        }
        const total = OVERHEAD + length;
        if (buf.length < total) return { frame: null, advance: 0 };
        const crcOffset = HEADER_SIZE + length;
        const crcActual =
            (buf[crcOffset] |
                (buf[crcOffset + 1] << 8) |
                (buf[crcOffset + 2] << 16) |
                (buf[crcOffset + 3] << 24)) >>> 0;
        const crcExpected = crc32(buf.subarray(2, crcOffset));
        if (crcActual !== crcExpected) {
            this.errors++;
            return { frame: null, advance: 1 };
        }
        const frame = {
            seq: buf[4],
            msgId: buf[5],
            payload: buf.slice(HEADER_SIZE, HEADER_SIZE + length),
        };
        return { frame, advance: total };
    }
}
